### ISSUE TITLE
Property with constrained generic type cannot be assigned to

### ISSUE THREAD

**Issue by mihaibirsan** [Microsoft/TypeScript#18937]
**TypeScript Version:**  2.5.3

**Code**

```ts
// A *self-contained* demonstration of the problem follows...
class Person<TPersonProps extends PersonProps = PersonProps> {

    props: TPersonProps;

    constructor(name: string) {
        const specificProps: PersonProps = {
            name
        };
        this.props = specificProps;
        this.props.name = name;
    }
}

interface PersonProps {
    name: string;
}
```

**Expected behavior:**
Compiles with no errors.

Note that line 10 doesn't throw an error, which is also an expected behavior.

Ideally I would be able to simply assign an object that matches the PersonProps interface:
```ts
    constructor(name: string) {
        this.props = {
            name
        };
    }
```

**Actual behavior:**
```
test.ts(9,9): error TS2322: Type 'PersonProps' is not assignable to type 'TPersonProps'.
```

**Comment by aluanhaddad** [comment]
That would be undesirable. The type parameter is constrained to be assignable to `PersonProps` and to default to `PersonProps` if not specified or inferred, but that does not make the reverse true. `PersonProps` is not assignable to the type parameter.

```ts
const p = new PersonProps<{name: string, location: {x: number, y: number}}>('Jake');
const p.props.location.x; // runtime error.
```

**Comment by mihaibirsan** [comment]
Ah, yes, I can see how this would break down.

---

In this case I'm wondering what the language support is for modeling and initializing components with extensible property types. It looks like a flaw in my design now. A way around it is to switch to composing the components instead of subclassing, but maybe someone can share a better way.

Here's a more complete React-like example of what I mean (and how I arrived at this error):
* Panel is a super-class that defines some behavior and (eg. `fullscreen`)
* EditorPanel is a specific type of Panel that has its own behavior (eg. `readonly`).

Ideally, I would initialize `this.state` in the parent class, instead of doing that in each of its subclasses. (And this is where my example breaks down if `readonly` isn't optional, as it happens to currently be.)

```ts
class Component<P, S> {
    props: P;
    state: S;

    constructor(props: P) {
        this.props = props;
    }
}

class Panel<TProps extends PanelProps = PanelProps, TState extends PanelState = PanelState>
        extends Component<TProps, TState> {

    constructor(props: TProps) {
        super(props);
        this.state = {
            fullscreen: false
        };
    }
}

interface PanelProps {
    name: string;
}

interface PanelState {
    fullscreen: boolean;
}

class EditorPanel extends Panel<EditorPanelProps> {

    constructor(props: EditorPanelProps) {
        super(props);
    }

    toString(): string {
        return JSON.stringify(this);
    }
}

interface EditorPanelProps extends PanelProps {
  readonly?: boolean;
}
```

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
